### DISCUSSION TITLE
在View层通过DataContext改变ViewModel中的属性或调用ViewModel中的方法是否违背MVVM呢？

### DISCUSSION THREAD

**Discussion by Case-Closed-X** [BYJRK/DotNet-Discussions#37] [MVVM]
在`View`层通过`DataContext`改变`ViewModel`中的属性或者调用`ViewModel`中的方法是否违背MVVM呢？

**Comment by BYJRK** [comment]
看你做到什么程度了。一般地，其实 ViewModel 能做的事情有限，经常需要借助附加属性、行为、触发器等方式来曲线救国。所以还有另外一种做法，也就是“不让 xaml.cs 文件完全空着”。比如说想要在程序加载后异步调用一个方法，常见操作是注册窗口的 Loaded 事件。那么可以考虑把 Loaded 方法改为 `async void`，然后在里面调用 VM 中的方法。这种其实没什么。但是如果把一些后台逻辑也直接或间接放在了 View 中，那么可能就违背了

**Reply by BYJRK** [reply]
此外，还可以用 Messenger 进行 View 与 ViewModel 之间的通信。如果只是想要进行一个简单的异步调用之类的，那么通常是没有问题的。

**Reply by Case-Closed-X** [reply]
好的，我其实还有一些疑问，就是直接在`ViewModel`中控制前台任何控件的`IsEnabled`或者`Visibility`属性违背MVVM吗？如果违背的话，是因为`ViewModel`中持有了`View`层的类吗？如果在`ViewModel`中控制除了按钮之外控件的`IsEnabled`属性，毕竟只是布尔值，也算违背MVVM吗？
还有就是如果我想在初始化`ObservableCollection`数据完毕之后，前台`ListBox`进行透明度淡入的动画，这个可以或者说只能通过附加属性、行为、触发器、自定义控件或者消息实现吗？

**Reply by KiyanYang** [reply]
说一下我的理解：

MVVM 在于 VM 和 V 的解耦，也就是说 VM 不知道 V 的情况，那 VM 肯定就不能持有 V 的类，而 `IsEnabled` 之类的属性，理论上说应该是 VM 自身的需求导致的，比如

```
public string UserName {get;set;}

public bool CanChangeUserName {get;set}
```

我们需要在某些时候禁止改变 `UserName`（也就是 `CanChangeUserName` 为 `False`），那么在 V 里就需要通过合适的方法来体验该属性，比如禁用提交按钮，或者禁止输入，或者隐藏输入。

而如果在设计 VM 的时候去过多考虑 V，很容易就会在 VM 里增加一个属性来控制某控件的外观或行为，我觉得这其实就算违反了 MVVM 思想。

V 调用 VM 里的方法，我觉得是可以的，之前我看过一篇文章，问 WPF 中 V 的 code-behind 里写了除了初始代码以外的代码是不是就不是 MVVM了，答：不一定， xaml 最后还是被编译为 c# code，所以 code-behind 里写代码一点问题也没，关键是写什么样的代码，一般情况下，只能是与 VM 输入和输出相关的代码，而不能去写 VM 的逻辑（比如写这样的 `vm.Prop1 = vm.Prop2 + vm.GetInc();` 这种 VM 的逻辑是 V 不能去写的）而且 V 自身也不应该包含过多业务逻辑。

说点题外话，一般都会说，MVVM 在于分离 VM 和 V，然后开发人员和设计人员可以协同工作，但是实际上界面很多都还还是开发人员写的😂。所以我觉得 MVVM 最重要的是极大简化了测试，不需要过多考虑 V，依靠 VM 就可以完成测试，那么在这种背景下，V 是控制台界面也无所谓，只有能与 VM 正确交互就可以了。

**Reply by Case-Closed-X** [reply]
切中肯綮，十分感谢二位的耐心解答
